### PR TITLE
ebooks

### PR DESCRIPTION
**Pull request by ghost**
book scan from camera menu (limited to books without chapters)
 sdcard needed for transfer
 eink recipes get updated when opening eink read menu (browsing recipes in tablet menu removed)
 read e-books on eink
 also find ebooks on memory cards (instead of random solitary recipes)
 scanning takes time, requires good lighting and consumes charges progressively, its also ~probably~ *definitely* suspendible
 books get identified during scanning (terse output)

#### Summary

SUMMARY: Feature "scan books with cameras and read them on eink"

#### Purpose of change

get rid of annoying clutter and provide obvious functionality

#### Describe the solution

re-purpose read inventory filter code, create dummy books with 0 weight/volume and start reading on returned location, remove all dummies after activity finishes/gets cancelled

#### Describe alternatives you've considered

refactoring visitable/containers, but my c++ is very rusty and im way to lazy for that

#### Additional context

scan times: 10min per 250ml book: so 30min for typical textbook up to 80min for a hefty volume like atomic survival
charge consumption: tool charges_to_use every 1min

played a bit with adding some 'boredom' morale penalty but eventually decided against it since twiddling your thumbs for two hours over food dehydrator or drilling in a wall carries none

### CODE REVIEW

**Comment by Inglonias** [comment]
I don't think you can digitize a book without more processing power than your average survivor has, or a LOT of spare time to transcribe the book by hand.

That said, finding ebooks on SD cards would be really neat.

File: src/iuse.cpp [hunk 1]
```diff
@@ -6751,6 +6834,68 @@ int iuse::camera( player *p, item *it, bool, const tripoint & )
...
+        mc.unset_flags();
+        mc.item_tags.insert( "MC_HAS_DATA" );
+        mc.set_var( "MC_BOOK", book.typeId() );
+        
+        // item copy( item( book.type, calendar::turn ) );
+        // copy.set_var( "weight", 0 );
+        // copy.item_tags.insert( "NO_DROP" );
+        // copy.item_tags.insert( "IRREMOVABLE" );
+        // mc.contents.erase();
+        // mc.put_in( copy );
```
Comment (Night-Pryanik) [inline src/iuse.cpp]: Please remove commented section if you don't need it.

**Comment by Night-Pryanik** [comment]
> I don't think you can digitize a book without more processing power than your average survivor has, or a LOT of spare time to transcribe the book by hand.

The books don't even need to be recognized by some OCR software, they could be shot by a tablet's camera and used as photos.

**Comment by dommon** [comment]
> > I don't think you can digitize a book without more processing power than your average survivor has, or a LOT of spare time to transcribe the book by hand.
> 
> The books don't even need to be recognized by some OCR software, they could be shot by a tablet's camera and used as photos.

true, but considering capabilities of modern quad-core arm7 based smartphone, and the logical conclusion that the tabled would have similar capabilities or more we can safely assume ocr'ing would not a problem either

cleanup commit coming, just compiling to test if all works as it should

**Comment by dissociativity** [comment]
Any ordinary smartphone newer than 2011 with a more or less basic app including open source ones can make a pdf that an ereader can access using photos or an app that intelligently scans text documents, it's hardly a reach, especially given we tend to start with smartphones now.

There's no real reason for a smartphone not to be able to do this, although it should take a long time proportional to book volume or length and drain lots of battery accordingly since having both the screen and camera on is fairly intense for any battery.

**Comment by dommon** [comment]
> If you can give the smartphone the abilities an e-ink tablet has, but with far less energy efficiency (10% an hour for reading actions, 20% an hour for anything using the camera or more processing power) as part of this PR, that would be excellent

too much coding effort for virtually no gain, eink is basically made for the job of being ereader and using it as such dosnt require changing reading checks (light level) or adding new ones (reading on tool ? pass tool used to activity and do power drain if applicable) to maintain logical consistency, also having smartphone double as inefficient ereader (or multiple ereader devices in general) would in effect necessitate adding a way to at least transfer ebooks between the devices when upgrading, possibly more intricate ways of managing their contents

**Comment by paulenka-aleh** [comment]
I suggest adding lighting and flat surface requirements for scan. Scanning on the floor would take more time and make your back ache. Scanning at poor lighting would require flash, multiple attempts to make a good shot and would drain your battery fast.

**Comment by egorzaderiy** [comment]
This PR will probably conflict with #30672

**Comment by Entity-y** [comment]
Instead of exclusively taking pictures of the contents of a book, what if books had a code that allowed access to a digital copy to be downloaded onto a device.

**Comment by kevingranade** [comment]
Yea sure, just as soon as the internet is back up.

**Comment by Entity-y** [comment]
It would make sense that internet services are still operational, even if access is limited, nobody is around to deactivate the service for their customers.

**Comment by kevingranade** [comment]
There's no electricity, there are no networks, there are no working broadcast towers, there are no web servers.  In short, no internet.

**Comment by Night-Pryanik** [comment]
> In short, no internet.

Off topic, but this reminds me that in case of no internet we need to justify working ATMs somehow.

**Comment by Entity-y** [comment]
I suppose you're right about all that. So is it out of the realm of possibilities that an enclave of civilization is still maintaining these faculties. Some of these buildings are already dotted around the procedurally generated world, once secured NPCs could be trained to keep these services going for the remaining survivors, some end game content to work towards, even an organized group like the refugee center could string something like this together. Better yet, a lab that hasn't succumbed to the catastrophic horrors above ground with actual scientists still roaming the halls, they could be that enclave.

**Comment by dommon** [comment]
ackchually, there is no reason for some limited version of area communications not working, satellites are designed completely maintenance free, any backbone relay network using them exclusively would be perfectly fine for at least a decade, likely two - so would anything with its own power supply using such network. not that big of a stretch to have working atms, in fact labs and mil installations etc should have contact with each other

**Comment by dommon** [comment]
> Scanning on the floor would take more time and make your back ache.

frankly, not unless youre some sort of boomer grandpa it wont, but as soon as there is helper function for crafting surface available - sure, why not
> Scanning at poor lighting would require flash, multiple attempts to make a good shot and would drain your battery fast.

indeed, ive added substantial charges requirement scaled by volume, possibly might add ongoing lighting check, but any further complexity is not enough bang for the buck

File: src/activity_handlers.cpp [hunk 2]
```diff
@@ -3805,6 +3812,16 @@ void activity_handlers::study_spell_finish( player_activity *act, player *p )
     }
 }
 
+void activity_handlers::scan_book_finish( player_activity *act, player *p )
+{
+    item &book = *act->targets[0];
+    item &mc = *act->targets[1];
+    mc.item_tags.insert( "MC_HAS_DATA" );
+    mc.set_var( "MC_BOOK", book.typeId() );
+    p->add_msg_if_player( _( "You meticulously scan %s and store it on sd card." ), book.tname() );
```
Comment (AMurkin) [inline src/activity_handlers.cpp]: I suggest call it everywhere either `SD-Memory card` (like item called), or `memory card` (to be more generic).

**Comment by davidpwbrown** [comment]
> ackchually, there is no reason for some limited version of area communications not working, satellites are designed completely maintenance free, any backbone relay network using them exclusively would be perfectly fine for at least a decade, likely two - so would anything with its own power supply using such network. not that big of a stretch to have working atms, in fact labs and mil installations etc should have contact with each other

Nah, most satellites would be dead within hours to weeks.
GPS would drift as ground stations didn't give updates anymore, positional tracking of other sats would then produce errors, satellites would either point the wrong way, or go into shutdown mode or not be able to change their orbit or other things depending on their configuration.

In short, any network relying on satellite comms would be either intermittently useless fairly soon after the apocalypse or completely useless within days to weeks.

The ATM issue has been discussed and other solutions have been found for them still working that fits with the lore, but this is kinda offtopic anyways.

**Comment by dommon** [comment]
> > ackchually, there is no reason for some limited version of area communications not working, satellites are designed completely maintenance free, any backbone relay network using them exclusively would be perfectly fine for at least a decade, likely two - so would anything with its own power supply using such network. not that big of a stretch to have working atms, in fact labs and mil installations etc should have contact with each other
> 
> Nah, most satellites would be dead within hours to weeks.
> GPS would drift as ground stations didn't give updates anymore, positional tracking of other sats would then produce errors, satellites would either point the wrong way, or go into shutdown mode or not be able to change their orbit or other things depending on their configuration.
> 
> In short, any network relying on satellite comms would be either intermittently useless fairly soon after the apocalypse or completely useless within days to weeks.
> 
> The ATM issue has been discussed and other solutions have been found for them still working that fits with the lore, but this is kinda offtopic anyways.

most, yes - but you would need only a few working reference stations with independent power supply to maintain dedicated network operational, and any reasonable military would have such contingencies in place
but its completely off topic, indeed

**Comment by Night-Pryanik** [comment]
> any reasonable military

They are zombified or dead.

**Comment by dommon** [comment]
>They are zombified or dead.

eventually, sure - but thats the thing tho, despite obvious warnings they apparently got completely caught with a hand in the chamber pot and no preparatory hardening of communication systems or failsafes went into effect, i find it easier to suspend disbelief for zombiefied michael jacksons million hp

**Comment by kevingranade** [comment]
> notable feature/bug - dummy books get created at every start of reading session, hence anything with chapters will always stay 'fresh'

This is not ok.

> scan times: 10min per 250ml book: so 30min for typical textbook up to 80min for a hefty volume like atomic survival

This sounds kind of short, does this have any kind of volume to page and time per page assumptions behind it?

Call it 600 pages for your typical textbook, 10s per pair of pages, 3,000s = 50 min

**Comment by dommon** [comment]
> > notable feature/bug - dummy books get created at every start of reading session, hence anything with chapters will always stay 'fresh'
> 
> This is not ok.

why? i dont know if i ever finished more than one 'for fun' book in any of my games ever (many 100+days, a few 1000+), and that one i did finish was when i deliberately tried to read 'for fun' more

there is a way to handle it (by moving book dummies in and out of eink) but it introduces significant performance penalty (crafting inventory and recipes code will go over that (possibly) hundreds of books inside eink every single time) for solving literal non issue

> 
> > scan times: 10min per 250ml book: so 30min for typical textbook up to 80min for a hefty volume like atomic survival
> 
> This sounds kind of short, does this have any kind of volume to page and time per page assumptions behind it?
> 
> Call it 600 pages for your typical textbook, 10s per pair of pages, 3,000s = 50 min
> 

650 pages of organic chemistry sitting on my desk has volume circa 1200ml, which is in line with game sizes for chemistry textbooks and id say around average size for rl science textbook. it really depends what you call typical tho, most of textbook group items have volume of 750ml
now, that was actually very, very conservative guesstimate from me, as even holding camera with one hand resting it on chest/shoulder and flipping pages with the other 5s is plenty, and that would give ~5min per 250ml

> Also you have json load errors

where

**Comment by dommon** [comment]
improvement on expanding on basic flavor text (You need to have a study book to scan.) would be welcome

**Comment by kevingranade** [comment]
> there is a way to handle it (by moving book dummies in and out of eink) but it introduces significant performance penalty

Find another way then, because neither feature breakage nor performance regression is acceptable.

> fun books can get created with 0 chapters left,

No changes to this feature are acceptable, no hacks.

> even holding camera with one hand resting it on chest/shoulder and flipping pages with the other 5s is plenty,

That kind of thing is not going to consistently yield readable images.

**Comment by Hymore246** [comment]
This PR would implement #29996

I'd recommend creating a designated book scanner item to do the scanning. It might be less practical than a cellphone but it would help you avoid all these "what-if" problem about lighting and photo quality.

**Comment by dommon** [comment]
> > there is a way to handle it (by moving book dummies in and out of eink) but it introduces significant performance penalty
> 
> Find another way then, because neither feature breakage nor performance regression is acceptable.
> 
> > fun books can get created with 0 chapters left,
> 
> No changes to this feature are acceptable, no hacks.

i cant even, well, partly my fault as i obviously meant that fun _E_-books would be created with 0 chapters on opening tablet read menu

its moot anyhow as i just gated scanning to books without chapters as was explained in commit message and updated pr description


> > even holding camera with one hand resting it on chest/shoulder and flipping pages with the other 5s is plenty,
> 
> That kind of thing is not going to consistently yield readable images.

debatable, but thats why i gave it ultra conservative 10min per 250ml which is exactly in line with your proposed scan times of 10s per (two) pages in the first place - to avoid this whole issue, silly me

**Comment by dommon** [comment]
> I'd recommend creating a designated book scanner item to do the scanning. It might be less practical than a cellphone but it would help you avoid all these "what-if" problem about lighting and photo quality.

there isnt any, lighting checks have been implemented and scan times are slow enough

**Comment by paulenka-aleh** [comment]
As one who scanned books and copybooks eith camera extensively during while being a student I can say that with a good lighting and flat surface you can quickly scan 2 pages of a copybook in 5-10 seconds (depending on binging or ability for someone else to hold it straight by the way, some bindings make photographing 2 pages at one impractical). Still it could produce poor-quality images, not all text gets focused all the time. Sometimes shadow over text or even flash might make content partially unreadable. If you check photos afterwards you spend a lot kore time, otherwise you might get hard time guessing what was the particular formula, definition or anything of a major imortance that can't be recognized on photo. Furthermore, unless you're reading sowly from PC (aka studying) photos of two pages become very inconvenient to read. Not all e-ink readers allow to scale such books which makes already poor-quality text even harder to read. On the phone you need to navigate pages both vertically and horizontally too and small screen might make viewing a single page hard. Unless you cut each page pair in two later. And since all pages are rotated a bit it's not a quick ask too.

10s per two pages sounds generous and I'd say could work only on excuse that e-ink tablet camera was specifically fesigned to scan boojs and that there's an app doing most of the hard work automatically except holding a book and page rotation...

**Comment by tinukedaya** [comment]
Personally if I would ever want to scan a book using a camera or a smartphone I would first build myself a little helper. Just a pipe/2x4 or similar with a holder for the camera and flashlight and maybe some frame/clamps to hold the book steady and straight.
 That way I have always the correct lightning, no shaking and the book always in the same distance and place.

So why not increase the freehand scanning time to even longer to abstract that some pages are being taken more then once because the first try was blurry and make it faster than that using some kind of tool?

╔══╦═╗
║---◙ /☼\\
║
║  ≡≡w≡≡
╚══════╛
_(I had to try this...)_

**Comment by I-am-Erk** [comment]
Closed due to inactivity. Please comment if you'd like it reopened.

**Pull request closed without merging**